gt/main.c

```diff
@@ -244,13 +247,164 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
 		"gt: receiving a packet with IP source address %s, and destination address %s, whose destination IP address is not the Grantor server itself.!\n",
 		src, dst);
 }
+static void
+gt_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	if (pcall_again == NULL) {
+		clear_ether_cache(eth_cache);
+		return;
+	}
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+	write_sequnlock(&eth_cache->lock);
+
+	*pcall_again = true;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ */
+static int
+gt_fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+	uint16_t inner_ip_ver, void *ip_dst, struct gatekeeper_if *iface)
+{
+	int ret;
+	unsigned lcore_id = rte_lcore_id();
+
+	eth_cache->stale = true;
+	eth_cache->ip_addr.proto = inner_ip_ver;
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		rte_memcpy(&eth_cache->ip_addr.ip.v4,
+			ip_dst, sizeof(eth_cache->ip_addr.ip.v4));
+	} else {
+		rte_memcpy(&eth_cache->ip_addr.ip.v6,
+			ip_dst, sizeof(eth_cache->ip_addr.ip.v6));
+	}
+
+	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(inner_ip_ver);
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+	rte_atomic32_set(&eth_cache->ref_cnt, 1);
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		ret = hold_arp(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	} else {
+		ret = hold_nd(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	}
+
+	if (ret < 0)
+		clear_ether_cache(eth_cache);
+
+	return ret;
+}
+
+static int
+drop_cache_entry_randomly(struct neighbor_hash_table *neigh)
+{
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	uint32_t entry_id = rte_rand() % neigh->tbl_size;
+
+	index = rte_hash_iterate(neigh->hash_table, &key, &data, &next);
+	while (index >= 0) {
+		int ret;
+		struct ether_cache *eth_cache;
+
+		if (entry_id != 0) {
+			index = rte_hash_iterate(neigh->hash_table,
+				&key, &data, &next);
+			entry_id--;
+			continue;
+		}
+
+		eth_cache = &neigh->cache_tbl[index];
+		ret = rte_hash_del_key(neigh->hash_table, key);
+		if (likely(ret >= 0))
+			clear_ether_cache(eth_cache);
+		else {
+			RTE_LOG(ERR, HASH,
+				"gk: failed to call rte_hash_del_key() at %s - %s",
+				__func__, strerror(-ret));
+		}
+
+		return ret;
+	}
+
+	return -1;
+}
+
+static struct ether_cache *
+gt_neigh_get_ether_cache(struct neighbor_hash_table *neigh,
+	uint16_t inner_ip_ver, void *ip_dst, struct gatekeeper_if *iface)
+{
+	int eth_idx;
+	struct ether_cache *eth_cache;
+	struct lls_config *lls_conf = get_lls_conf();
```

Comment:
Postpone the call to `get_lls_conf()` to after the lookup and test below. Thus, we leave the minimum amount of work possible on the hot path, namely the lookup and test.